### ISSUE TITLE
How to test Dates with timezone?

### ISSUE THREAD

**Issue by shred** [joel-costigliola/assertj-core#1123]
I'm sorry for opening an issue, but I have found no examples for this problem, and I'm not sure if it is actually a missing feature.

#### Summary

Let's say I have to unit test a `Date` object, like this one:

```java
Calendar cal = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
cal.clear();
cal.set(2017, Calendar.NOVEMBER, 27, 20, 13, 24);
Date date = cal.getTime();
```

If I understood the documentation right, I would have to assert it like this:

```java
assertThat(date).isEqualTo("2017-11-27T21:13:24");
```

However, this unit test has a problem. It runs fine on my system with `Europe/Berlin` timezone. But if the same test is run on a different machine with different timezone settings, it would fail.

Besides that, it isn't clearly readable because I expect the time to be "20:13:24" (UTC), but have to compare it against "21:13:24" (CET).

I have tried this:

```java
assertThat(date)
    .withDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'")
    .isEqualTo("2017-11-27T20:13:24Z");
```

But the test fails, because it still expects `"2017-11-27T21:13:24Z"`.

Giving an explicit time zone in the equal to string fails for the same reason:

```java
assertThat(date)
    .withDateFormat("yyyy-MM-dd'T'HH:mm:ssZ")
    .isEqualTo("2017-11-27T20:13:24+00:00");
```

I have found no way to give `AbstractDateAssert` a timezone. How I am supposed to write unit tests that do not depend on the system's timezone?

I am using JDK 7 and assertj 2.8.0. The unit test is for a library, so I cannot switch to JDK 8 yet, to keep the requirements simple.

Thank you!

**Comment by joel-costigliola** [comment]
The [isEqualTo](http://joel-costigliola.github.io/assertj/core/api/org/assertj/core/api/AbstractDateAssert.html#isEqualTo(java.lang.String)) date assertion javadoc states that the the default formats are expressed in the current local timezone. `withDateFormat` is just a way to register another date format to be used in the following assertions.

All the assertion methods taking `String` parameters are just convenience to avoid users to parse the expected dates, it that does not work as you expect, I suggest you to use assertion methods taking `Date` parameters where you have full control of everything.

The second example fails but AssertJ does not do anything special, it just converts the `String` into a `Date` from the `SimpleDateFormat` built with `"yyyy-MM-dd'T'HH:mm:ss'Z'"` and then compare it to actual. I do agree it is confusing but I'm not sure what AssertJ can do better in that case (I'm open to suggestions).

The last example actually fails to parse the given String is not a correct date for `"yyyy-MM-dd'T'HH:mm:ssZ"`, you can run that:
```java
new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssZ").parse("2017-11-27T20:13:24+00:00");
```
AssertJ the falls back to use default data formats which happen to be able to parse the String, that behavior should be clearly stated in the javadoc, that's something to improve on our side.

What improvements would like to see in the API ? Feedback is always welcome for us.

**Comment by shred** [comment]
Thank you for your explanation! Now I understand how the date asserts works.

I also found out that I made a mistake in my example above:

```java
assertThat(date)
    .withDateFormat("yyyy-MM-dd'T'HH:mm:ssZ")
    .isEqualTo("2017-11-27T20:13:24+00:00");  // <-- wrong timezone format
```

Actually, `Z` cannot parse `+00:00`, but only `+0000`. I have to use `X` for testing ISO 8601 formatted dates. The following assert now works as expected, and uses the given time zone:

```java
assertThat(date)
    .withDateFormat("yyyy-MM-dd'T'HH:mm:ssX")
    .isEqualTo("2017-11-27T20:13:24+00:00");
```

It also works if I add
```java
AbstractDateAssert.registerCustomDateFormat("yyyy-MM-dd'T'HH:mm:ssX");
```
to the top of the unit test, and don't give individual `withDateFormat()`.

Maybe assertj should add the `"yyyy-MM-dd'T'HH:mm:ssX"` format to its set of default date formats?
